ArchiSteamFarm/ArchiWebHandler.cs

```diff
@@ -1131,19 +1131,39 @@ internal sealed class ArchiWebHandler : IDisposable {
 
 		internal void OnVanityURLChanged(string vanityURL = null) => VanityURL = string.IsNullOrEmpty(vanityURL) ? null : vanityURL;
 
-		internal async Task<(EResult Result, EPurchaseResultDetail? PurchaseResult)?> RedeemWalletKey(string key) {
+		internal async Task<(EResult Result, EPurchaseResultDetail PurchaseResult)> RedeemWalletKey(string key) {
 			if (string.IsNullOrEmpty(key)) {
 				Bot.ArchiLogger.LogNullError(nameof(key));
-				return null;
+				return (EResult.Timeout, EPurchaseResultDetail.Timeout);
 			}
 
-			const string request = "/account/validatewalletcode";
+			const string requestValidateCode = "/account/validatewalletcode";
 
 			// Extra entry for sessionID
 			Dictionary<string, string> data = new Dictionary<string, string>(2) { { "wallet_code", key } };
 
-			Steam.RedeemWalletResponse response = await UrlPostToJsonObjectWithSession<Steam.RedeemWalletResponse>(SteamStoreURL, request, data).ConfigureAwait(false);
-			return response == null ? ((EResult Result, EPurchaseResultDetail? PurchaseResult)?) null : (response.Result, response.PurchaseResultDetail);
+			Steam.RedeemWalletResponse responseValidateCode = await UrlPostToJsonObjectWithSession<Steam.RedeemWalletResponse>(SteamStoreURL, requestValidateCode, data).ConfigureAwait(false);
+			if (responseValidateCode == null) {
+				return (EResult.Timeout, EPurchaseResultDetail.Timeout);
+			}
+
+			if (responseValidateCode.Result != EResult.OK) {
+				return (responseValidateCode.Result, responseValidateCode.PurchaseResultDetail);
+			}
+
+			const string requestCheckFunds = "/account/createwalletandcheckfunds";
```

Comment:
This is required only on accounts without wallet created, and in its current form is awfully wrong as being executed on already-walleted accounts. This should be its own function and available only when `EResult` explicitly states that the wallet is not yet available, not by definition in all conditions.

Reply:
And if it really changed (which I heavily doubt), at the minimum you should record whole traffic (for people with wallets and without) in order to spot a difference (because there should be something) and add proper logic for that.

Reply:
Current code doesn't work at all, I have checked this too. Will check their JS scripts soon

Reply:
For reference: https://steamstore-a.akamaihd.net/public/javascript/redeemwalletcode.js?v=OsL8Jt8oANiA

Reply:
From what I see and based on my gut with working with Steam's crap, just executing `confirmredeemwalletcode` should do the trick for walleted accounts, while those without wallet would need to execute `createwalletandcheckfunds` firstly, and since that one might require extra details (such as city, country, address), it probably shouldn't be implemented by ASF.

It'd probably work if instead of whole this PR you've just changed the request address. Valve probably also moved it there, and re-implemented old endpoint as extra validation (but not redemption). It's interesting, since it allows to check if codes are valid without redeeming them 🤔

Reply:
I asked my friend and he said that they use validate endpoint for determining whether user `haswallet` so the JS code knows whether it should to create wallet or redeem wallet code next.

VALVE COME ON YOU REALLY COULD DO THAT BETTER.

Idea for now: skip whole `validate` mess, check if user has wallet via our recent `Bot` additions, if he doesn't have wallet during redeem, throw the error right there (keeping original "can't activate from client" would do), if he has wallet call `confirmredeemwalletcode` which is current implementation, just with changed request address.

If the above solution works, this is most likely what ASF should stick with. ASF should not create wallets on its own, and the AWH endpoint should not be executed with no wallet available, to not run into the error to begin with. The remaining bit (once this is confirmed working) is to check whether `detail` is always available in new endpoint (valid/invalid/dupe key), and if yes, update JSON definition accordingly.

Reply:
Also createwalletandcheckfunds is necessary when wallet and wallet code have different currencies, so can it be executed in that case?

Reply:
> ASF should not create wallets on its own (...) since that one might require extra details (such as city, country, address)

Reply:
And detail is not always available (in case such as invalid sessionid)

Reply:
It doesn't create wallet in that case, it just converting wallet code currency to already existing account wallet currency

Reply:
Yes, I've checked the code and it looks like you could create wallet (omitting address/city/everything else) but only if user already has existing wallet available.

So in this case, proper order is:
a) When AH finds out about wallet key, check if user already has wallet
b) If not, keep original error and do not enter AWH at all
c) If yes, call that crappy `validatewalletcode` and do the same as JS:

```
		if ( result.haswallet )
		{
			if ( result.currency == result.wallet.currencycode )
			{
								RedeemWalletCode();	
			}
			else
			{
								CreateWalletAndCheckFunds( false );
			}
		}
```

d) Notice how create wallet response decides whether to create new wallet in given currency or convert, you should use the same logic and send `confirm` only if currency matches:

```
		if ( result.currency != result.grant_currency )
		{
			$('code_amount').innerHTML = g_sWalletCodeAmount;
			$('code_exchange_amount').innerHTML = g_sExchangedWalletCodeAmount;

			DisplayPage( 'redeem_amount' );
		}
		else
		{
						RedeemWalletCode();						
		}
```

e) If no new wallet is needed, just call `RedeemWalletCode`

Reply:
You need proper JSON structures for all above details, also with proper nullable types if the endpoint is not returning some field (e.g. due to invalid session). AWH should also have 3 functions for this, while logic should be somewhere in `Actions` probably.

Reply:
That redeem_amount page redeems code as well, so we can not to check it at all

Reply:
> unless that page redeems wallet code as well further below

Yup.